service/mapper/transitions.go

```diff
@@ -349,9 +353,24 @@ func (t *Transitions) IndexChain(s *State) error {
 		if err != nil {
 			return fmt.Errorf("could not index header: %w", err)
 		}
-		blockID := header.ID()
+		blockID = header.ID()
 		log = log.Hex("block", blockID[:])
 	}
+	if t.cfg.IndexHeight {
+		// If we don't already have the block ID, retrieve it now.
+		if blockID == flow.ZeroID {
+			header, err := t.chain.Header(s.height)
+			if err != nil {
+				return fmt.Errorf("could not get header: %w", err)
+			}
+			blockID = header.ID()
+			log = log.Hex("block", blockID[:])
+		}
```

Comment:
Adding this amount of complexity for a small optimization that is most likely completely irrelevant is definitely not worth it.